### DISCUSSION TITLE
MOU run health check

### DISCUSSION THREAD

**Discussion by laat0003** [gmdsi/GMDSI_notebooks#110] [Q&A] [answered]
Hi,

MOU runs be taking long times.

Is there anything I should be looking out for that could indicate things have gone awry.  Or is it a case of "if it is running then that means its working"?

Cheers,
Tariq

**Comment by mnfienen** [comment]
Hey Tariq

Those MOU runs take really forever - good to run it over night. So yeah - if it's running it's working.

Cheers,
Mike

**Reply by laat0003** [reply]
Apologies I should have been clearer with my question. I meant MOU runs in general.  I tend to work from an assumption that I've done something wrong and consequently expect things to fail. Because MOU runs require a huge computational burden, knowing what to look out for during the optimization process could save lots of time if things are not progressing as they should.

**Comment by rhugman** [comment, marked as answer]
@laat0003  I will usually have a "parallel" script/notebook/whatever that reads MOU's outputs at each generation to keep an eye on progress. Can sometimes be useful to catch #dumbmistakes early on (and retain your sanity).

pestpp-mou outputs lots of info (so many files!) at each generation; this + rec file will often give plenty of diagnostic info.

**Reply by laat0003** [reply]
Thanks @rhugman. I'm keen on the specifics of the #dumbmistakes you mention and how they might be spotted in the outputs. That said, a great addition to these notebooks could be an "errors to watch for when using XXX" section akin to the end of chapters in "Applied GW modelling". I actually wanted to ask Cecile about her failures in our session today because they're usually better to learn from (for me anyway) than the published successes - but I hog question time too much so just typed it and deleted it three or four times.

**Reply by rhugman** [reply]
#dumbmistakes can be many (unpredictable) things...like setting the wrong sign on the one-sided constraints, or maximizing instead of minimizing..."user error" types of things. I'll usually be plotting up the pareto front as it progresses through the generations, things like that. Making sure the numbers make sense kind of thing.

Im sure others have better/more systematic approaches. I tend to design on the fly based on the problem at hand.

Good thing we can get here to respond here then :) @Cecile-A-C

**Reply by jtwhite79** [reply]
@laat0003 I think @rhugman 's approach to plotting the front(s) as mou evolves is probably the most robust...but if you are lazy (like me), you might also be able to get away with watching the decision variable, constraint, and objective summary stats - if those stats have stabilized (especially the min and max of each objective), then mou might be converged - thats why mou reports changes in these each generation.  But plotting/viz is probably the clearest way to see what's going on across generations.

@rqmacasieb also pointed out that we can watch more proper objective characteristics like the "knee" of the objective tradeoff (the solution nearest the optimal tradeoff) and things related to solution crowding/hypervolume as ways to monitor convergence so watch for those coming out soon!  

if you are using constraints, one of the things that will def trip mou up is a (nearly) infeasible population.  mou will continuing trying to find feasible solutions for many generations, but if the number of feasible members isnt improving, you'll probably have to ctrl+c...

**Reply by laat0003** [reply]
Thanks @jtwhite79. I agree that visual cues from plotting are great for keeping an eye on things. When I did pareto analysis with vanilla PEST I mostly focused on the knee and assumed that with MOU it would be equally informative.

**Comment by Cecile-A-C** [comment]
Hi @laat0003, I had a lot of checks for OPT but not as many for MOU. Usually I looked at the rec file and I tried plotting everything I could from the outputs files.

For OPT, I looked at the objective function, decision variables and constraints vs iteration (in my case I found that sequential linear programming gave much better results than linear programming, and I could also detect situations where OPT wasn't converging - I would see decision variables/constraints oscillating from one iteration to another). It took me some testing to find feasible sets of initial decision variables and a good decision variable increment. With chance constraints, I looked at the constraint PDFs associated with the optimal decision variables, and I checked that the probability of constraint violation corresponded to the "risk" value (e.g. with a risk (i.e. reliability) of 0.95, the expected probability of constraint violation would be 5%).

For MOU, I looked at the Pareto front vs generation (in my case it was very long to try to get the extremities of the Pareto front to converge). On my first tries I didn't have any feasible individuals in the initial decision variable population, so I had to use pretty narrow bounds when generating that initial dvar population. When I used risk as an objective, I picked a few solutions from the final Pareto front and checked if the probability of constraint violation corresponded to the "risk" value (and that wasn't always the case - I probably hadn't fully converged to the Pareto front yet).

**Reply by laat0003** [reply]
Hi @Cecile-A-C ,

Thanks for the informative reply. I'm playing around with an MOU approach on a project at the moment, so the information is timely. It's also nice to know that my struggles with OPT when I used it for the first time are common. I too hit a wall finding feasible sets of initial decision variables and really got stuck on a suitable increment. I don't get the opportunity to use either of these optimization utilities very often which is a shame because I really enjoy working with them. When I do get a chance to use them, I feel like I have to learn everything from scratch and often forget many of the checks and/or nuances around problem configuration, which makes all the replies in this Q&A very useful moving forward.
